fdbserver/DDTeamCollection.actor.cpp

```diff
@@ -1554,17 +1590,26 @@ class DDTeamCollectionImpl {
 					when(wait(server->ssVersionTooFarBehind.onChange())) {}
 					when(wait(self->disableFailingLaggingServers.onChange())) {}
 					when(wait(server->longStorageQueue.onChange())) {
-						TraceEvent(SevInfo, "TriggerStorageQueueRebalance", self->distributorId)
-						    .detail("SSID", server->getId());
-						std::vector<ShardsAffectedByTeamFailure::Team> teams;
-						for (const auto& team : server->getTeams()) {
-							std::vector<UID> servers;
-							for (const auto& server : team->getServers()) {
-								servers.push_back(server->getId());
+						Optional<int64_t> threshold = selectLongStorageQueueThreshold(self->teams);
+						// threshold represents the queue length of majority SSes
+						if (!threshold.present() || server->longStorageQueue.get() < threshold.get()) {
+							TraceEvent(SevInfo, "TriggerStorageQueueRebalanceIgnored", self->distributorId)
+							    .detail("ThresholdPresent", threshold.present())
+							    .detail("SSID", server->getId());
+						} else {
+							TraceEvent(SevInfo, "TriggerStorageQueueRebalance", self->distributorId)
+							    .detail("SSID", server->getId());
+							std::vector<ShardsAffectedByTeamFailure::Team> teams;
+							for (const auto& team : server->getTeams()) {
```

Comment:
The longest storage queue is calculated per `longest queue within team`, and the data move trigger is based on individual server's queue length, would this make the reasoning complex?

Reply:
I think the reasoning here is not complicated --- even more straightforward. If a SS queue is very long, then: (1) we avoid selecting this SS to move in the data; (2) we select a shard to move out of this SS. To achieve (1), we avoid selecting any team which has this SS to move in the data. To achieve (2), we go throughput all shards owned by this SS and select one to move out. In concept, we care about servers; In implementation, we use teams to operate shards.

Reply:
I think the percentage calculation will be complex, we are using the longest queue from each team, this will hide the servers with short queue, the calculated percentage is not straightforward, it is not the actual percentage?

Reply:
Yes. The percentage calculation is an estimation. We use the percentage of team to estimate the percentage of servers. 

"this will hide the servers with short queue" 
Suppose we have an unbalanced team which has two short queues and one long queue, we still want to avoid moving shards into the team and want to move shards out of the team, since the server with a long queue is a cluster-wide blocker. With our rebalance policy, we move shards out of this hot server and make the queue size small for all servers in the team. This is even helpful to rebalance the queue size of servers within an unbalanced team.